### PR TITLE
[IMP] payment term improvement

### PR DESCRIPTION
**Pull request by hmo-odoo**
Task: https://www.odoo.com/web#id=19773&view_type=form&model=project.task&menu_id=3942&action=327

Pad: https://pad.odoo.com/p/r.qInm1aE9zon5mNOJ

### CODE REVIEW

File: addons/account/models/account_invoice.py [hunk 1]
```diff
@@ -1222,12 +1222,19 @@ def compute(self, value, date_ref=False):
             elif line.value == 'balance':
                 amt = round(amount, prec)
             if amt:
-                next_date = (datetime.strptime(date_ref, '%Y-%m-%d') + relativedelta(days=line.days))
-                if line.days2 < 0:
+                next_date = datetime.strptime(date_ref, '%Y-%m-%d')
```
Comment (hmo-odoo) [inline addons/account/models/account_invoice.py]: use fields.date.from_string method

see: https://github.com/odoo/odoo/blob/master/openerp/fields.py#L1264

File: addons/account/models/account_invoice.py [hunk 2]
```diff
@@ -1222,12 +1222,19 @@ def compute(self, value, date_ref=False):
...
-                    next_date += relativedelta(day=line.days2, months=1)
+                    next_date = next_first_date + relativedelta(days=line.days-1)
+                elif line.option == 'fix_day_current_month':
+                    next_first_date = next_date + relativedelta(day=1, months=0)  # Getting 1st of current month
+                    next_date = next_first_date + relativedelta(days=line.days-1)
+                elif line.option == 'last_day_following_month':
+                    next_date += relativedelta(day=31, months=1)  # Getting last day of next month
+                elif line.option == 'last_day_current_month':
+                    next_date += relativedelta(day=31, months=0)  # Getting last day of next month
                 result.append((next_date.strftime('%Y-%m-%d'), amt))
```
Comment (hmo-odoo) [inline addons/account/models/account_invoice.py]: use fields.date.to_string()

File: addons/account/views/account_view.xml [hunk 3]
```diff
@@ -1321,10 +1321,11 @@
             <field name="model">account.payment.term.line</field>
             <field name="arch" type="xml">
                 <tree string="Payment Term">
-                    <field name="value"/>
+                    <field name="sequence" widget="handle"/>
+                    <field name="value" string="Due Type"/>
                     <field name="value_amount" attrs="{'readonly':[('value','=','balance')]}"/>
                     <field name="days"/>
-                    <field name="days2"/>
+                    <field name="option" string=""/>
```
Comment (hmo-odoo) [inline addons/account/views/account_view.xml]: why none title (string) ?

File: addons/account/views/account_view.xml [hunk 4]
```diff
@@ -1343,8 +1344,11 @@
                             </div>
                         </group>
                         <group string="Due Date Computation">
-                            <field name="days"/>
-                            <field name="days2"/>
+                            <label for="days"/>
+                            <div>
+                                <field name="days" class="oe_inline" attrs="{'invisible': [('option', 'in', ['last_day_following_month', 'last_day_current_month'])]}"/>
```
Comment (hmo-odoo) [inline addons/account/views/account_view.xml]: @ged-odoo ,

is it ok to use oe_inline class here ?

File: addons/account/views/account_view.xml (hunk 3 above)
Comment (mdi-odoo) [inline addons/account/views/account_view.xml]: It's due to usability improvement.

File: addons/account/models/account_invoice.py [hunk 5]
```diff
@@ -1171,17 +1171,25 @@ class AccountPaymentTerm(models.Model):
...
     company_id = fields.Many2one('res.company', string='Company', required=True, default=lambda self: self.env.user.company_id)
 
     @api.constrains('line_ids')
     @api.one
     def _check_lines(self):
-        if self.line_ids and self.line_ids[-1].value != 'balance':
-            raise UserError("A Payment Term should have its last line of type Balance")
+        payment_term_lines = self.line_ids.sorted()
+        if payment_term_lines and payment_term_lines[-1].value != 'balance':
+            raise ValidationError('A Payment Term should have its last line of type Balance')
```
Comment (hmo-odoo) [inline addons/account/models/account_invoice.py]: error message string should be translatable.

File: addons/account/models/account_invoice.py [hunk 6]
```diff
@@ -1171,17 +1171,25 @@ class AccountPaymentTerm(models.Model):
...
+
+        lines = self.line_ids.filtered(lambda r: r.value == 'balance')
+        if len(lines) > 1:
+            raise ValidationError('A Payment Term should have only one line of type Balance')
```
Comment (hmo-odoo) [inline addons/account/models/account_invoice.py]: see above comment

File: addons/account/views/account_view.xml (hunk 4 above)
Comment (ged-odoo) [inline addons/account/views/account_view.xml]: no

Comment (ged-odoo) [inline addons/account/views/account_view.xml]: add a class o_row on the outside div, and remove the oe_inline classes

**Comment by qdp-odoo** [comment]
merged in revision 6e5db46

thanks

**Pull request closed without merging**